dit-core/src/daemon.rs

```diff
@@ -1 +1,158 @@
+// manage the remote peers that connect to the listener.
 
+use crate::codec::Codec;
+use crate::peer::PeerConfig;
+use futures_util::{SinkExt, StreamExt};
+use serde::{Deserialize, Serialize};
+use std::net::SocketAddr;
+use tokio::io;
+use tokio::net::{TcpListener, TcpStream};
+use tokio_util::codec::Framed;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Config {
+    pub peer_config: PeerConfig,
+    pub daemon_config: DaemonConfig,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct DaemonConfig {
+    pub socket_addr: SocketAddr,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum Packet {
+    Ping(u64),
+    Pong(u64),
+}
+
+#[derive(Debug)]
+/// This struct represents the connection from the daemon to a process.
+pub struct ConnectionToProcess {
+    stream: Framed<TcpStream, Codec<Packet>>,
+}
+
+impl ConnectionToProcess {
+    /// Create a new outbound connection to a process.
+    pub fn new(socket: TcpStream) -> Self {
+        let stream = Framed::new(socket, Codec::new(1024));
+        Self { stream }
+    }
+
+    /// Run the inbound connection decoding packets from the stream and
+    /// responding to them.
+    pub async fn run(mut self) -> io::Result<()> {
+        tracing::info!("Running inbound connection");
+
+        // Get packets from stream
+        while let Some(packet) = self.stream.next().await {
+            match packet {
+                Ok(Packet::Ping(value)) => {
+                    tracing::info!("Received Ping with value: {}", value);
+
+                    // Respond with Pong
+                    self.stream.send(Packet::Pong(value)).await?;
+                }
+                Ok(Packet::Pong(_)) => {
+                    tracing::warn!("Received unexpected Pong packet");
+                }
+                Err(e) => {
+                    tracing::error!("Failed to process packet: {}", e);
+                    return Err(e);
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+/// This struct represents the connection to the daemon from a process.
+pub struct ConnectionToDaemon {
+    stream: Framed<TcpStream, Codec<Packet>>,
+}
+
+impl ConnectionToDaemon {
+    /// Create a new connection to the daemon.
+    pub async fn connect(address: SocketAddr) -> tokio::io::Result<Self> {
+        let stream = TcpStream::connect(address).await?;
+        let stream = Framed::new(stream, Codec::new(1024));
+
+        Ok(Self { stream })
+    }
+
+    /// Send a ping packet to the daemon.
+    pub async fn ping(&mut self, value: u64) -> tokio::io::Result<()> {
+        self.stream.send(Packet::Ping(value)).await
+    }
+
+    /// Receive a packet from the daemon.
+    pub async fn receive(&mut self) -> tokio::io::Result<Option<Packet>> {
+        match self.stream.next().await {
+            Some(Ok(packet)) => Ok(Some(packet)),
+            Some(Err(e)) => Err(e),
+            None => Ok(None),
+        }
+    }
```

Comment:
Es wäre vermutlich besser "ping senden + pong empfangen" als eine Funktion zu haben und nicht dem Nutzer der Bibliothek nicht die "rohen" Pakete in die Hand zu geben. Das musst du aber nicht in dieser PR machen.